### PR TITLE
Log a warning when extraneous parameters are passed into a request

### PR DESCRIPTION
**Pull request by tsmacdonald**
Cherry-picked in @camsaul 's logging updates from #27688 for my own ergonomics.

image

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/28032)

### CODE REVIEW

File: src/metabase/api/common.clj [hunk 1]
```diff
@@ -261,16 +262,49 @@
...
   {:pre [(or (string? route) (vector? route))]}
-  `(def ~(vary-meta fn-name
-                    assoc
-                    :doc          docstr
-                    :is-endpoint? true)
-     (~(symbol "compojure.core" (name method)) ~route ~args
-      ~@body)))
+  (let [method-kw      (method-symbol->keyword method)
+        allowed-params (keys arg->schema)
+        prep-route     #'compojure/prepare-route]
```
Comment (tsmacdonald) [inline src/metabase/api/common.clj:292]: doesn't feel great grabbing that private fn, but despite being private it seems unlikely to change. And if it ever does the entire app will stop working (and most tests will fail) and we'll know instantly.

File: src/metabase/api/common.clj [hunk 2]
```diff
@@ -261,16 +262,50 @@
                                   (ns-name *ns*) fn-name)))
       (assoc parsed :fn-name fn-name, :route route, :docstr docstr))))
 
+(defn validate-param-values
```
Comment (metamben) [inline src/metabase/api/common.clj:265]: Looking at this name, I expect it to throw an exception if the param values are not valid.

File: src/metabase/api/common.clj (hunk 1 above)
Comment (metamben) [reply src/metabase/api/common.clj:293]: Or it changes subtly, so as to break production only. 😁

File: src/metabase/api/common.clj [hunk 3]
```diff
@@ -261,16 +262,50 @@
...
+  "Log a warning if the request body contains any parameters not included in `expected-params` (which is presumably
+  populated by the defendpoint schema)"
+  [{route :compojure/route body :body} expected-params]
+  (when (map? body)
+    (let [extraneous-params (set/difference (set (keys body))
+                                            (set expected-params))]
+      (when (seq extraneous-params)
+        (log/warn (format (str "Unexpected parameters at %s: %s\nPlease add them to the schema or remove them from the"
```
Comment (camsaul) [inline src/metabase/api/common.clj:273]: You could use `log/warnf` which combines `log`/warn` + `format`

It's always a good idea to add a little note saying why you're not i18n'ing something. If this is mean to be dev-facing only then that's fine, just add a note to that effect so people don't think you simply forgot to use `trs`. On that note: **should** this only be enabled in `:dev` mode? Suppose the FE client is sending a key and we accept it but don't have a schema to validate it: should we be filling our user's logs with an untranslated warning they can't do anything about?

**Review by camsaul — APPROVED** [review]
Overall LGTM but you should consider making that error message be `:dev`-only

File: src/metabase/api/common.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/api/common.clj:273]: I guess maybe you could do it in `:dev` + `:test` mode but not `:prod` so your test still works

**Review by metamben — APPROVED** [review]
LGTM

**Pull request merged**